### PR TITLE
User registration

### LINKED ISSUE #1: User Registration
For a user to register they must provide their:

1. Name
2. Email
3. Password

---

When a user registers, they get a wallet created for them

### PR DESCRIPTION
**Pull request by aldemeery**
This PR (Pull Request) adds the user registration feature.

Closes #1

### CODE REVIEW

File: app/Http/Controllers/Auth.php [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Illuminate\Http\Request;
+
+class Auth extends Controller
```
Comment (aldemeery) [inline app/Http/Controllers/Auth.php]: @aaldemeery This controller is not being used, please delete it

File: app/Http/Controllers/AuthController.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+class AuthController extends Controller
+{
+    public function register()
+    {
+        return view('auth/register');
+    }
+
+    public function signup(Request $request)
+    {
+
```
Comment (aldemeery) [inline app/Http/Controllers/AuthController.php]: @aaldemeery Remove extra empty lines

File: app/Http/Controllers/AuthController.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+        $data = $request->validate([
+            'name' => 'required|string',
```
Comment (aldemeery) [inline app/Http/Controllers/AuthController.php]: @aaldemeery Name should not be longer than 255 characters

File: app/Http/Controllers/AuthController.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            'email' => 'required|email|unique:users,email',
```
Comment (aldemeery) [inline app/Http/Controllers/AuthController.php]: @aaldemeery Same for email, no longer than 255 chars

File: app/Http/Controllers/AuthController.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        $data = $repuest->validate([
+            'email' => ['required', 'email'],
+            'password' => ['required'],
+        ]);
+
+        if (Auth::attempt($data)) {
+            request()->session()->regenerate();
+
+            dd('succeful');
```
Comment (aldemeery) [inline app/Http/Controllers/AuthController.php]: @aaldemeery Using `dd` terminates everything, you need to `return` something instead

File: app/Http/Controllers/CategoryController.php [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Illuminate\Http\Request;
+
+class CategoryController extends Controller
```
Comment (aldemeery) [inline app/Http/Controllers/CategoryController.php]: @aaldemeery Remove the empty controller

File: app/Http/Controllers/ImageController.php [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Illuminate\Http\Request;
+
+class ImageController extends Controller
```
Comment (aldemeery) [inline app/Http/Controllers/ImageController.php]: @aaldemeery Same here

File: app/Http/Controllers/ProductController.php [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Illuminate\Http\Request;
+
+class ProductController extends Controller
```
Comment (aldemeery) [inline app/Http/Controllers/ProductController.php]: @aaldemeery And here

File: app/Http/Controllers/ReviewController.php [hunk 9]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Illuminate\Http\Request;
+
+class ReviewController extends Controller
```
Comment (aldemeery) [inline app/Http/Controllers/ReviewController.php]: @aaldemeery And here

File: app/Http/Controllers/StoreController.php [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Illuminate\Http\Request;
+
+class StoreController extends Controller
```
Comment (aldemeery) [inline app/Http/Controllers/StoreController.php]: @aaldemeery And here

File: app/Http/Controllers/VoucherController.php [hunk 11]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Illuminate\Http\Request;
+
+class VoucherController extends Controller
```
Comment (aldemeery) [inline app/Http/Controllers/VoucherController.php]: @aaldemeery And here

File: app/Http/Controllers/WalletController.php [hunk 12]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Illuminate\Http\Request;
+
+class WalletController extends Controller
```
Comment (aldemeery) [inline app/Http/Controllers/WalletController.php]: @aaldemeery And here

File: app/Models/Category.php [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+    protected $fillable = [
+        "name",
+    ];
+
+    public function products(): HasMany
+    {
+        return $this ->hasMany(Product::class);
+    }
+
```
Comment (aldemeery) [inline app/Models/Category.php]: @aaldemeery Remove the extra empty line

File: app/Models/Category.php [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+{
+    use HasFactory;
+
+    protected $fillable = [
+        "name",
+    ];
+
+    public function products(): HasMany
+    {
+        return $this ->hasMany(Product::class);
```
Comment (aldemeery) [inline app/Models/Category.php]: @aaldemeery Remove the extra empty space before the arrow

File: app/Models/Image.php [hunk 15]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace App\Models;
+
+use App\Models\Product;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+
+class image extends Model
```
Comment (aldemeery) [inline app/Models/Image.php]: @aaldemeery A class name must begin with an uppercase letter

File: app/Models/Product.php [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+use App\Models\Category;
+use Illuminate\Database\Eloquent\Model;
+use App\Http\Controllers\ReviewController;
+use Illuminate\Database\Eloquent\Relations\HasMany;
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+
+class Product extends Model
+{
+    use HasFactory;
```
Comment (aldemeery) [inline app/Models/Product.php:17]: @aaldemeery Add an empty line after the use statement, and fix your indentation, you're using 3 spaces instead of 4

File: app/Models/Product.php [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+     protected $fillable = [
+        'name',
```
Comment (aldemeery) [inline app/Models/Product.php:20]: @aaldemeery Add `about` to the fillable

File: app/Models/Product.php [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+        'price',
+        'quantity',
+     ];
+
+     protected $casts = [
+        "price" => "integer",
```
Comment (aldemeery) [inline app/Models/Product.php]: @aldemeery A price should be cast into a float not integer

File: app/Models/Review.php [hunk 19]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace App\Models;
+
+use App\Models\Store;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+
+class review extends Model
```
Comment (aldemeery) [inline app/Models/Review.php]: @aaldemeery A class name MUST begin with an uppercase letter

File: app/Models/Review.php [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+{
+    use HasFactory;
+
+    public function review_id()
```
Comment (aldemeery) [inline app/Models/Review.php]: @aaldemeery Do not add methods before attributes...take this down below the `$fillable` array

File: app/Models/Review.php [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        return $this->morphTo();
+    }
+
+    protected $fillable = [
+        "title",
+        "content"
```
Comment (aldemeery) [inline app/Models/Review.php]: @aaldemeery Add the stars to the fillable

File: app/Models/User.php [hunk 22]
```diff
@@ -38,7 +42,14 @@ class User extends Authenticatable
      *
      * @var array<string, string>
      */
-    protected $casts = [
-        'email_verified_at' => 'datetime',
```
Comment (aldemeery) [inline app/Models/User.php:42]: @aaldemeery Why did you remove the `$casts` array from here? Please add it back

File: app/Models/Voucher.php [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+
+class Voucher extends Model
+{
+    use HasFactory;
+
+    protected $fillable = [
+        "percentage",
+        "remaning",
```
Comment (aldemeery) [inline app/Models/Voucher.php]: @aaldemeery You have a typo in `remaning`, it should be `remaining` instead

File: app/Models/Voucher.php [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+        "code",
+    ];
+
```
Comment (aldemeery) [inline app/Models/Voucher.php:19]: @aaldemeery Add a `$casts` array

File: app/Models/Wallet.php [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+class Wallet extends Model
+{
+    use HasFactory;
+
+    protected $fillable = [
+        "balance",
+    ];
+
+    protected $casts = [
+        "balance" => "decimal"
```
Comment (aldemeery) [inline app/Models/Wallet.php]: @aaldemeery Change `decimal` to `float`

File: database/factories/CategoryFactory.php [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+class CategoryFactory extends Factory
+{
+    /**
+     * Define the model's default state.
+     *
+     * @return array<string, mixed>
+     */
+    public function definition()
+    {
+        return [
```
Comment (aldemeery) [inline database/factories/CategoryFactory.php]: @aaldemeery All your factories have empty definitions, either add them (the definitions) or just remove the factory classes altogether for now

File: database/migrations/2014_10_12_000000_create_users_table.php [hunk 27]
```diff
@@ -17,9 +17,7 @@ public function up()
             $table->id();
             $table->string('name');
             $table->string('email')->unique();
-            $table->timestamp('email_verified_at')->nullable();
```
Comment (aldemeery) [inline database/migrations/2014_10_12_000000_create_users_table.php:20]: @aaldemeery Do not remove these, we need them both

File: database/migrations/2022_11_16_210900_create_categories_table.php [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('categories', function (Blueprint $table) {
+            $table->id();
+            $table->string('name');
```
Comment (aldemeery) [inline database/migrations/2022_11_16_210900_create_categories_table.php]: @aaldemeery The name is unique

File: database/migrations/2022_11_16_210935_create_products_table.php [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('products', function (Blueprint $table) {
+            $table->id();
+            $table->foreignId('category_id')->constrained()->onUpdate('cascade')->onDelete('cascade');
+            $table->foreignId('store_id')->constrained()->onUpdate('cascade')->onDelete('cascade');
+            $table->string('name');
```
Comment (aldemeery) [inline database/migrations/2022_11_16_210935_create_products_table.php:20]: @aaldemeery Add an `about` column

File: database/migrations/2022_11_16_211112_create_wallets_table.php [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('wallets', function (Blueprint $table) {
+            $table->id();
+            $table->foreignId('user_id')->constrained()->onUpdate('cascade')->onDelete('cascade');
```
Comment (aldemeery) [inline database/migrations/2022_11_16_211112_create_wallets_table.php]: @aaldemeery Since the user can have only one wallet, the `user_id` needs to be unique

File: routes/web.php [hunk 31]
```diff
@@ -16,3 +17,8 @@
 Route::get('/', function () {
     return view('welcome');
 });
+
+Route::get('register', [AuthController::class, 'register'])->name('register');
+Route::post('signup', [AuthController::class, 'signup'])->name('singup');
+Route::get('login', [AuthController::class, 'login'])->name('signin');
+Route::post('signin', [AuthController::class, 'signin'])->name('signin');
```
Comment (aldemeery) [inline routes/web.php]: @aaldemeery These are two routes with different paths, yet with the same name!! Change the name

**Comment by aaldemeery** [comment]
@aldemeery Changes implemented, please re-review

**Comment by aaldemeery** [comment]
user registeration
